replace SIGNAL and SLOT with functors in the audio engine

Be-ing opened the pull request:
This will make it easier to improve tangled code by making ControlObject connections easier to find with IDEs.

Be-ing committed:
CO::connectValueChangeRequest replace SLOT with functors

Be-ing committed:
LoopingControl: replace SIGNAL & SLOT with functors

rryan commented on src/engine/sync/internalclock.cpp:
```diff
@@ -36,7 +36,7 @@ InternalClock::InternalClock(const char* pGroup, SyncableListener* pEngineSync)
     m_pSyncMasterEnabled->setButtonMode(ControlPushButton::TOGGLE);
     m_pSyncMasterEnabled->connectValueChangeRequest(
-        this, SLOT(slotSyncMasterEnabledChangeRequest(double)),
+        this, [=](double value){slotSyncMasterEnabledChangeRequest(value);},
```
Why a lambda here (and in other diffs) instead of `&InternalClock::slotSyncMasterEnabledChangeRequest`?

Be-ing replied:
Because that doesn't compile:
```
src/engine/loopingcontrol.cpp: In constructor ‘LoopingControl::LoopingControl(QString, UserSettingsPointer)’:
src/engine/loopingcontrol.cpp:131:81: error: no matching function for call to ‘ControlObject::connectValueChangeRequest(LoopingControl*, void (LoopingControl::*)(double), Qt::ConnectionType)’
          &LoopingControl::slotBeatLoopSizeChangeRequest, Qt::DirectConnection);
                                                                              ^

In file included from src/engine/loopingcontrol.cpp:7:
src/control/controlobject.h:154:10: note: candidate: ‘bool ControlObject::connectValueChangeRequest(const QObject*, std::function<void(double)>, Qt::ConnectionType)’
     bool connectValueChangeRequest(const QObject* receiver,
          ^~~~~~~~~~~~~~~~~~~~~~~~~
src/control/controlobject.h:154:10: note:   no known conversion for argument 2 from ‘void (LoopingControl::*)(double)’ to ‘std::function<void(double)>’
```

Be-ing replied:
Lambdas seem to be the easiest way to use a class member function as a std::function.

Be-ing replied:
std::bind also works, but IMO the `_1` placeholder thing it requires is weird and I'd prefer to use a lambda.

rryan replied:
Ah that's because you're making connectValueChanged take a std::function, while pointer to member is its own type and not convertible to std::function. 

You can get around this by making connectValueChanged a template so you aren't writing down the type of method, then you can pass through method to  connect() and the appropriate version of connect will get called based on whether it's pointer to member or a functor.

Be-ing committed:
ControlProxy::connectValueChanged support Qt 5 functor syntax

Be-ing committed:
BpmControl: replace SIGNAL & SLOT with Qt 5 functor syntax

Be-ing committed:
ClockControl: replace SIGNAL & SLOT with Qt 5 functor syntax

Be-ing committed:
CueControl: replace SIGNAL & SLOT with Qt 5 functor syntax

daschuer commented on src/mixer/playermanager.cpp:
```diff
@@ -64,15 +64,15 @@ PlayerManager::PlayerManager(UserSettingsPointer pConfig,
         m_pTrackAnalysisScheduler(TrackAnalysisScheduler::NullPointer()) {
     m_pCONumDecks->connectValueChangeRequest(this,
-            SLOT(slotChangeNumDecks(double)), Qt::DirectConnection);
+            [=](double value){slotChangeNumDecks(value);}, Qt::DirectConnection);
```
We should enable here the normal member function pointer syntax. 
This is the Qt solution: 
```
    template <typename Func1, typename Func2>
    static inline QMetaObject::Connection connect(const typename QtPrivate::FunctionPointer<Func1>::Object *sender, Func1 signal,
                                     const typename QtPrivate::FunctionPointer<Func2>::Object *receiver, Func2 slot,
                                     Qt::ConnectionType type = Qt::AutoConnection)
```

I think it is something like 

```
template <typename Func>
bool ControlDoublePrivate::connectValueChangeRequest(
        const typename QtPrivate::FunctionPointer<Func>::Object *receiver, 
        Func slot, 
        Qt::ConnectionType type) {
```

Be-ing replied:
That is a better solution, thank you.

Be-ing replied:
I tried this but I have not gotten it to work:
```
In file included from src/track/keyutils.h:8,
                 from src/track/trackrecord.h:7,
                 from src/track/track.h:10,
                 from src/engine/cachingreader.h:16,
                 from src/engine/enginebuffer.h:25,
                 from src/engine/bpmcontrol.cpp:7:
src/control/controlproxy.h: In instantiation of ‘bool ControlProxy::connectValueChanged(const QObject*, Func, Qt::ConnectionType) [with Func = void (BpmControl::*)(double)]’:
src/control/controlproxy.h:97:58:   required from ‘bool ControlProxy::connectValueChanged(Func, Qt::ConnectionType) [with Func = void (BpmControl::*)(double)]’
src/engine/bpmcontrol.cpp:42:60:   required from here
src/control/controlproxy.h:79:13: error: no matching function for call to ‘ControlProxy::connect(ControlProxy*, void (ControlProxy::*)(double), const QObject*&, void (BpmControl::*&)(double), Qt::ConnectionType&)’
         if (!connect(this, &ControlProxy::valueChanged, receiver, method, scoConnection)) {
In file included from /usr/include/qt5/QtCore/QObject:1,
                 from src/control/controlobject.h:21,
                 from src/engine/bpmcontrol.cpp:3:
/usr/include/qt5/QtCore/qobject.h:228:43: note: candidate: ‘static QMetaObject::Connection QObject::connect(const typename QtPrivate::FunctionPointer<Func>::Object*, Func1, const typename QtPrivate::FunctionPointer<Func2>::Object*, Func2, Qt::ConnectionType) [with Func1 = void (ControlProxy::*)(double); Func2 = void (BpmControl::*)(double); typename QtPrivate::FunctionPointer<Func>::Object = ControlProxy; typename QtPrivate::FunctionPointer<Func2>::Object = BpmControl]’ <near match>
     static inline QMetaObject::Connection connect(const typename QtPrivate::FunctionPointer<Func1>::Object *sender, Func1 signal,
                                           ^~~~~~~
/usr/include/qt5/QtCore/qobject.h:228:43: note:   conversion of argument 3 would be ill-formed:
In file included from src/track/keyutils.h:8,
                 from src/track/trackrecord.h:7,
                 from src/track/track.h:10,
                 from src/engine/cachingreader.h:16,
                 from src/engine/enginebuffer.h:25,
                 from src/engine/bpmcontrol.cpp:7:
src/control/controlproxy.h:79:57: error: invalid conversion from ‘const QObject*’ to ‘const Object*’ {aka ‘const BpmControl*’} [-fpermissive]
         if (!connect(this, &ControlProxy::valueChanged, receiver, method, scoConnection)) {
                                                         ^~~~~~~~
In file included from /usr/include/qt5/QtCore/QObject:1,
                 from src/control/controlobject.h:21,
                 from src/engine/bpmcontrol.cpp:3:
/usr/include/qt5/QtCore/qobject.h:208:36: note: candidate: ‘static QMetaObject::Connection QObject::connect(const QObject*, const char*, const QObject*, const char*, Qt::ConnectionType)’
     static QMetaObject::Connection connect(const QObject *sender, const char *signal,
                                    ^~~~~~~
/usr/include/qt5/QtCore/qobject.h:208:36: note:   no known conversion for argument 2 from ‘void (ControlProxy::*)(double)’ to ‘const char*’
/usr/include/qt5/QtCore/qobject.h:211:36: note: candidate: ‘static QMetaObject::Connection QObject::connect(const QObject*, const QMetaMethod&, const QObject*, const QMetaMethod&, Qt::ConnectionType)’
     static QMetaObject::Connection connect(const QObject *sender, const QMetaMethod &signal,
                                    ^~~~~~~
/usr/include/qt5/QtCore/qobject.h:211:36: note:   no known conversion for argument 2 from ‘void (ControlProxy::*)(double)’ to ‘const QMetaMethod&’
/usr/include/qt5/QtCore/qobject.h:463:32: note: candidate: ‘QMetaObject::Connection QObject::connect(const QObject*, const char*, const char*, Qt::ConnectionType) const’
 inline QMetaObject::Connection QObject::connect(const QObject *asender, const char *asignal,
                                ^~~~~~~
/usr/include/qt5/QtCore/qobject.h:463:32: note:   candidate expects 4 arguments, 5 provided
/usr/include/qt5/QtCore/qobject.h:260:13: note: candidate: ‘template<class Func1, class Func2> static typename std::enable_if<((int)(QtPrivate::FunctionPointer<Func2>::ArgumentCount) >= 0), QMetaObject::Connection>::type QObject::connect(const typename QtPrivate::FunctionPointer<Func>::Object*, Func1, Func2)’
             connect(const typename QtPrivate::FunctionPointer<Func1>::Object *sender, Func1 signal, Func2 slot)
             ^~~~~~~
/usr/include/qt5/QtCore/qobject.h:260:13: note:   template argument deduction/substitution failed:
In file included from src/track/keyutils.h:8,
                 from src/track/trackrecord.h:7,
                 from src/track/track.h:10,
                 from src/engine/cachingreader.h:16,
                 from src/engine/enginebuffer.h:25,
                 from src/engine/bpmcontrol.cpp:7:
src/control/controlproxy.h:79:13: note:   candidate expects 3 arguments, 5 provided
         if (!connect(this, &ControlProxy::valueChanged, receiver, method, scoConnection)) {
In file included from /usr/include/qt5/QtCore/QObject:1,
                 from src/control/controlobject.h:21,
                 from src/engine/bpmcontrol.cpp:3:
/usr/include/qt5/QtCore/qobject.h:269:13: note: candidate: ‘template<class Func1, class Func2> static typename std::enable_if<(((int)(QtPrivate::FunctionPointer<Func2>::ArgumentCount) >= 0) && (! QtPrivate::FunctionPointer<Func2>::IsPointerToMemberFunction)), QMetaObject::Connection>::type QObject::connect(const typename QtPrivate::FunctionPointer<Func>::Object*, Func1, const QObject*, Func2, Qt::ConnectionType)’
             connect(const typename QtPrivate::FunctionPointer<Func1>::Object *sender, Func1 signal, const QObject *context, Func2 slot,
             ^~~~~~~
/usr/include/qt5/QtCore/qobject.h:269:13: note:   template argument deduction/substitution failed:
/usr/include/qt5/QtCore/qobject.h: In substitution of ‘template<class Func1, class Func2> static typename std::enable_if<(((int)(QtPrivate::FunctionPointer<Func2>::ArgumentCount) >= 0) && (! QtPrivate::FunctionPointer<Func2>::IsPointerToMemberFunction)), QMetaObject::Connection>::type QObject::connect(const typename QtPrivate::FunctionPointer<Func>::Object*, Func1, const QObject*, Func2, Qt::ConnectionType) [with Func1 = void (ControlProxy::*)(double); Func2 = void (BpmControl::*)(double)]’:
src/control/controlproxy.h:79:13:   required from ‘bool ControlProxy::connectValueChanged(const QObject*, Func, Qt::ConnectionType) [with Func = void (BpmControl::*)(double)]’
src/control/controlproxy.h:97:58:   required from ‘bool ControlProxy::connectValueChanged(Func, Qt::ConnectionType) [with Func = void (BpmControl::*)(double)]’
src/engine/bpmcontrol.cpp:42:60:   required from here
/usr/include/qt5/QtCore/qobject.h:269:13: error: no type named ‘type’ in ‘struct std::enable_if<false, QMetaObject::Connection>’
src/control/controlproxy.h: In instantiation of ‘bool ControlProxy::connectValueChanged(const QObject*, Func, Qt::ConnectionType) [with Func = void (BpmControl::*)(double)]’:
src/control/controlproxy.h:97:58:   required from ‘bool ControlProxy::connectValueChanged(Func, Qt::ConnectionType) [with Func = void (BpmControl::*)(double)]’
src/engine/bpmcontrol.cpp:42:60:   required from here
/usr/include/qt5/QtCore/qobject.h:300:13: note: candidate: ‘template<class Func1, class Func2> static typename std::enable_if<(QtPrivate::FunctionPointer<Func2>::ArgumentCount == -1), QMetaObject::Connection>::type QObject::connect(const typename QtPrivate::FunctionPointer<Func>::Object*, Func1, Func2)’
             connect(const typename QtPrivate::FunctionPointer<Func1>::Object *sender, Func1 signal, Func2 slot)
             ^~~~~~~
/usr/include/qt5/QtCore/qobject.h:300:13: note:   template argument deduction/substitution failed:
In file included from src/track/keyutils.h:8,
                 from src/track/trackrecord.h:7,
                 from src/track/track.h:10,
                 from src/engine/cachingreader.h:16,
                 from src/engine/enginebuffer.h:25,
                 from src/engine/bpmcontrol.cpp:7:
src/control/controlproxy.h:79:13: note:   candidate expects 3 arguments, 5 provided
         if (!connect(this, &ControlProxy::valueChanged, receiver, method, scoConnection)) {
In file included from /usr/include/qt5/QtCore/QObject:1,
                 from src/control/controlobject.h:21,
                 from src/engine/bpmcontrol.cpp:3:
/usr/include/qt5/QtCore/qobject.h:308:13: note: candidate: ‘template<class Func1, class Func2> static typename std::enable_if<(QtPrivate::FunctionPointer<Func2>::ArgumentCount == -1), QMetaObject::Connection>::type QObject::connect(const typename QtPrivate::FunctionPointer<Func>::Object*, Func1, const QObject*, Func2, Qt::ConnectionType)’
             connect(const typename QtPrivate::FunctionPointer<Func1>::Object *sender, Func1 signal, const QObject *context, Func2 slot,
             ^~~~~~~
/usr/include/qt5/QtCore/qobject.h:308:13: note:   template argument deduction/substitution failed:
/usr/include/qt5/QtCore/qobject.h: In substitution of ‘template<class Func1, class Func2> static typename std::enable_if<(QtPrivate::FunctionPointer<Func2>::ArgumentCount == -1), QMetaObject::Connection>::type QObject::connect(const typename QtPrivate::FunctionPointer<Func>::Object*, Func1, const QObject*, Func2, Qt::ConnectionType) [with Func1 = void (ControlProxy::*)(double); Func2 = void (BpmControl::*)(double)]’:
src/control/controlproxy.h:79:13:   required from ‘bool ControlProxy::connectValueChanged(const QObject*, Func, Qt::ConnectionType) [with Func = void (BpmControl::*)(double)]’
src/control/controlproxy.h:97:58:   required from ‘bool ControlProxy::connectValueChanged(Func, Qt::ConnectionType) [with Func = void (BpmControl::*)(double)]’
src/engine/bpmcontrol.cpp:42:60:   required from here
/usr/include/qt5/QtCore/qobject.h:308:13: error: no type named ‘type’ in ‘struct std::enable_if<false, QMetaObject::Connection>’
```

Be-ing committed:
remove need for lambdas with connectValueChange(Request)

and remove old SLOT syntax for
ControlProxy::connectValueChangeRequest. That was easier than
getting it to play nice with the templating required for the
new functor syntax.

Be-ing committed:
QuantizeControl: replace SIGNAL & SLOT with Qt 5 functor syntax

Be-ing committed:
RateControl: replace SIGNAL & SLOT with Qt 5 functor syntax

Be-ing committed:
VinylControlControl: replace SIGNAL/SLOT with Qt 5 functor syntax

Be-ing committed:
move EngineFilter classes to their own folder

daschuer commented:
Did you get a notification about this: https://github.com/Be-ing/mixxx/pull/20

daschuer commented on src/control/control.h:
```diff
@@ -115,8 +115,14 @@ class ControlDoublePrivate : public QObject {
-    bool connectValueChangeRequest(const QObject* receiver,
-                                   const char* method, Qt::ConnectionType type);
+    template <typename Receiver, typename Slot>
+    bool connectValueChangeRequest(Receiver receiver,
```
This way there might be less type savety.
Does the compiler complain if slot is not part of the receiver?

Be-ing replied:
Yes, it will not compile if the slot is not part of the receiver.

daschuer commented on src/control/controlproxy.h:
```diff
@@ -30,10 +30,72 @@ class ControlProxy : public QObject {
+    // TODO: get the compiler to accept parent() for the receiver
+    //template <typename Slot>
+    //bool connectValueChanged(Slot func, Qt::ConnectionType type = Qt::AutoConnection) {
+    //    DEBUG_ASSERT(parent() != nullptr);
```
See https://github.com/Be-ing/mixxx/pull/20

Be-ing replied:
Oh, I did not see the notification about this earlier. However, I would rather continue with how it is here. I am really reluctant to put anything in Mixxx code that uses something called "QtPrivate". That could easily break without warning with a new Qt version.

rryan replied:
I think we could make our own template that does the same thing (if Func is a pointer to member Obj::Func, extract Obj as Object) pretty easily:

```c++
template<typename Func> struct FunctionPointer {};

template<class Obj, typename Ret, typename... Args> struct FunctionPointer<Ret (Obj::*) (Args...)> {
    typedef Obj Object;
};
template<class Obj, typename Ret, typename... Args> struct FunctionPointer<Ret (Obj::*) (Args...) const> {
    typedef Obj Object;
};
```

daschuer commented:
This makro is already used in the public interface of QObject in qobject.h so it is not too private.
Hover we have the choice to copy the implementation now or do it later if compilation fails with a Qt update.

Be-ing commented:
The name "QtPrivate" to me suggests that its presence in a public header is only an implementation detail which could change at any time. It is not mentioned in Qt documentation at all. I don't think it's a good idea to use in Mixxx code.

Be-ing commented:
I already went the hard route and replaced use of the shortcut function throughout Mixxx with the longer form where the receiver is explicitly specified. At this point I would rather keep that than create a situation where builds could fail with a Qt update (which could happen to a distro packager before we update our build environments with a new Qt version).

Thank you for sending me the PR though. Please leave a comment on the upstream PR when you do that (at least for me). I look at Zulip for GitHub notifications, so only see notifications for mixxxdj/mixxx unless I look at my email.

rryan commented on src/engine/loopingcontrol.cpp:
```diff
@@ -122,84 +122,84 @@ LoopingControl::LoopingControl(QString group,
-    connect(m_pCOBeatLoop, SIGNAL(valueChanged(double)), this,
-            SLOT(slotBeatLoop(double)), Qt::DirectConnection);
+    connect(m_pCOBeatLoop, &ControlObject::valueChanged, this,
+            [=](double value){slotBeatLoop(value);}, Qt::DirectConnection);
```
This one is still a lambda

Be-ing replied:
This is required because the slot takes more (default) parameters than the signal emits.

Be-ing committed:
move EngineControls to their own folder

daschuer commented:
I am fine with the short hand to remove the function which is using parent() though the solution requires an upcast via qobject_cast, which has runtime overhead.

Be-ing commented:
ControlProxy::connectValueChanged is not used very frequently (mostly at initialization), so I am not concerned about runtime overhead.

Be-ing committed:
Merge remote-tracking branch 'upstream/master' into engine_function_pointer_connections

daschuer commented:
> ControlProxy::connectValueChanged is not used very frequently (mostly at initialization), so I am not concerned about runtime overhead.

The runtime overhead is hidden insid the proposed qobject_cast. You solution is just an additional load and passing this, which should be fast.

Be-ing committed:
move EngineChannel classes to their own folder

Be-ing committed:
move CachingReader classes to their own folder

Be-ing committed:
move EngineBufferScaler classes to their own folder

Be-ing committed:
InternalClock: replace SIGNAL & SLOT with Qt 5 functor syntax

Be-ing committed:
EngineChannel: replace SIGNAL & SLOT with Qt 5 functor syntax

Be-ing committed:
EngineDeck: replace SIGNAL & SLOT with Qt 5 functor syntax

Be-ing committed:
CachingReader: replace SIGNAL & SLOT with Qt 5 functor syntax

Be-ing committed:
EngineDelay: replace SIGNAL & SLOT with Qt 5 functor syntax

Be-ing committed:
EngineTalkoverDucking: replace SIGNAL & SLOT with Qt 5 functors

Be-ing committed:
EngineBuffer: replace SIGNAL & SLOT with Qt 5 functor syntax

Be-ing commented:
Done. There are no instances of SIGNAL or SLOT macros left in src/engine :)

daschuer reviewed (commented):
Ups ... I have accidentally merged this from my local master branch after test. 
But not a big issue since it works good. 
I had only two comments. @Be-ing can you prepare a follow up?

daschuer commented on src/vinylcontrol/vinylcontrolmanager.cpp:
```diff
@@ -88,7 +88,7 @@ void VinylControlManager::slotNumDecksChanged(double dNumDecks) {
         ControlProxy* pEnabled = new ControlProxy(group, "vinylcontrol_enabled", this);
         m_pVcEnabled.push_back(pEnabled);
-        pEnabled->connectValueChanged(&m_vinylControlEnabledMapper, SLOT(map()));
+        pEnabled->connectValueChanged(&m_vinylControlEnabledMapper, QOverload<int>::of(&QSignalMapper::mapped));
```
But it is successful on Appveyor. Is the bug fixed or a false positive?

rryan replied:
I think that's about `qOverload`, which chooses the right overload from QOverload, QConstOverload, and QNonConstOverload.

http://doc.qt.io/qt-5/qtglobal.html#qOverload

daschuer commented:
Thank you for all that assembly-line work :-)

Be-ing commented:
Thank you for review. I would like to continue this with all Track loading and related signals soon to attempt to find the source of audio glitches on track load.

daschuer commented:
Go ahead, thank you.
